### DISCUSSION TITLE
Unable to create 'working' logical component

### DISCUSSION THREAD

**Discussion by wilbert-alberts** [labs4capella/python4capella#123] [Q&A]
Hi, 

Not sure whether this is a problem or a question. In my first steps with python4capella I wanted to create a script that creates a 'simulation' counterpart for each logical component in the model. The logic is shown below (based on 'List_logical_components_in_console.py'):
```
for lc in se.get_logical_architecture().get_logical_component_pkg().get_owned_logical_components():
    # : :type lc: LogicalComponent
    simSystems = []
    for l2 in lc.get_owned_logical_components():
        simComp = LogicalComponent()
        simComp.set_name('sim_' + l2.get_name())
        simComp.set_description('Simulation component of ' + l2.get_name())
        simSystems.append(simComp)
        
    model.start_transaction()
    lc.get_owned_logical_components().add_all(simSystems)
    model.commit_transaction()
```

This works in so far that it creates logical components. However, I wasn't able to drag them on a diagram in my model. In order to find any differences between the original logical components and the new ones, I checked the 'Expert' Properties. I found a difference in 'Abstract Typed Elements'. Did I do something wrong in creating the logical components?

The original one: 

![image](https://user-images.githubusercontent.com/4869849/164648729-4e1834a6-38bc-481f-8721-1eec8f426139.png)

The simulation one:

![image](https://user-images.githubusercontent.com/4869849/164648781-9752d023-c4ff-4948-8439-5d58c74ad46b.png)

Any help is apprecated.

Regards,
Wilbert

**Comment by ylussaud** [comment]
Some things are missing in the API to achieve what you want to do. In the Java world we have a method to copy EObject (elements from the Capella model), but it's not available in the Python API. I opened [this issue](https://github.com/labs4capella/python4capella/issues/124) to track this.
In your case it would be easier to use such API and then change the name of the LogicalComponent and set it's description. But since we don't have is yet you can use something like this as a workaround:

    simComp.get_java_object().getAbstractTypedElements().addAll(l2.get_java_object().getAbstractTypedElements())

That basically calls Java methods using the referenced Java Object of each Python Object. That's not the goal of Python4Capella since we aim at simplifying the API but it should help you until we add the copy() method.

**Reply by wilbert-alberts** [reply]
I see what you mean but `simComp.get_java_object().getAbstractTypedElements()` seems to return an UnmodiableEList so addAll is not allowed there...

**Reply by ylussaud** [reply]
It's a derived EReference... I'm not sure how it is computed but while debugging it seems to have something to do with a Part Object referencing the LogicalComponent. Maybe someone can help you on the [Capella forum](https://forum.mbse-capella.org/).
In this case you will have to copy more than one Object probably the Part object to.

**Reply by wilbert-alberts** [reply]
Sorry, I've been busy with some other tasks lately. 
I went a bit deeper in this and see what you mean with 'Part'. After opening the model with the sample reflective ecore modeler, I noticed that a Cappella recognizes the notion of a type and an instance. When I create a LogicalComponent like:

`simComp = LogicalComponent()`

I actually only create the type LogicalComponent. I also need to create a Part and tell it that it has the simComp as type. Then I can drag the simComp on the canvas. 

I don't know whether the distinction between type and instance becomes visible in Capella, but if not, then creating anything using constructors like `LogicalComponent()`, `PhysicalComponent()` etc, should create a corresponding Part as well I think. However, I've not figured out yet how to create a Part using the provided APIs so far. I created the Part within the reflective ecore editor.

**Comment by ylussaud** [comment]
We don't have API for creating Part... you might be able to use:

    create_e_object("http://www.polarsys.org/capella/core/cs/" + capella_version(), "Part")

To crate a direct reference to a Java EObject in you script. For the moment we lack a proper API to really modify the Capella model.
